### ISSUE TITLE
`tsc` execution speed improvements

### ISSUE THREAD

**Issue by valarnin** [quisquous/cactbot#4508]
## Summary

There are a number of code paths that slow `tsc` execution.

Steps taken to check:

1. `npm i -g @typescript/analyze-trace`
2. `npm run tsc-no-emit -- --generateTrace ../cactbot-tracing`
3. `npx analyze-trace ../cactbot-tracing --force-millis 300`

Results:

```
Hot Spots
├─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/ui/raidboss/popup-text.ts (2467ms)
│  └─ Check expression from (line 695, char 70) to (line 697, char 11) (1918ms)
│     └─ Check expression from (line 697, char 11) to (line 697, char 10) (1915ms)
├─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/ui/raidboss/data/05-shb/eureka/delubrum_reginae_savage.ts (721ms)
│  └─ Check variable declaration from (line 95, char 7) to (line 3679, char 2) (633ms)
│     └─ Check expression from (line 95, char 38) to (line 3679, char 2) (509ms)
│        └─ Check expression from (line 174, char 13) to (line 2567, char 4) (326ms)
├─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/ui/raidboss/data/05-shb/ultimate/the_epic_of_alexander.ts (573ms)
│  └─ Check variable declaration from (line 352, char 7) to (line 3191, char 2) (505ms)
│     └─ Check expression from (line 352, char 38) to (line 3191, char 2) (375ms)
│        └─ Check expression from (line 553, char 13) to (line 2701, char 4) (308ms)
├─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/test/helper/test_trigger.ts (540ms)
│  └─ Check deferred node from (line 391, char 34) to (line 562, char 4) (485ms)
│     └─ Check variable declaration from (line 520, char 20) to (line 520, char 28) (407ms)
│        └─ Check expression from (line 520, char 32) to (line 520, char 50) (407ms)
├─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/ui/raidboss/data/05-shb/eureka/zadnor.ts (451ms)
│  └─ Check variable declaration from (line 105, char 7) to (line 2316, char 2) (425ms)
│     └─ Check expression from (line 105, char 38) to (line 2316, char 2) (324ms)
│        └─ Check expression from (line 109, char 13) to (line 1659, char 4) (262ms)
├─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/ui/raidboss/data/04-sb/ultimate/unending_coil_ultimate.ts (434ms)
│  └─ Check variable declaration from (line 175, char 7) to (line 2036, char 2) (374ms)
│     └─ Check expression from (line 175, char 38) to (line 2036, char 2) (270ms)
│        └─ Check expression from (line 219, char 13) to (line 1575, char 4) (231ms)
├─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/ui/raidboss/data/05-shb/raid/e12s.ts (370ms)
│  └─ Check variable declaration from (line 278, char 7) to (line 2063, char 2) (297ms)
│     └─ Check expression from (line 278, char 38) to (line 2063, char 2) (207ms)
│        └─ Check expression from (line 281, char 13) to (line 1674, char 4) (179ms)
└─ Check file /mnt/c/Users/valarnin/Source/valarnin-cactbot/ui/raidboss/data/05-shb/eureka/delubrum_reginae.ts (315ms)
   └─ Check variable declaration from (line 43, char 7) to (line 1918, char 2) (291ms)
      └─ Check expression from (line 43, char 38) to (line 1918, char 2) (223ms)
         └─ Check expression from (line 46, char 13) to (line 1275, char 4) (159ms)

Duplicate packages
├─ schema-utils
│  ├─ Version 4.0.0 from /mnt/c/Users/valarnin/Source/valarnin-cactbot/node_modules/copy-webpack-plugin/node_modules/schema-utils
│  ├─ Version 4.0.0 from /mnt/c/Users/valarnin/Source/valarnin-cactbot/node_modules/webpack-dev-server/node_modules/schema-utils
│  └─ Version 3.1.1 from /mnt/c/Users/valarnin/Source/valarnin-cactbot/node_modules/webpack/node_modules/schema-utils
└─ ajv
   ├─ Version 6.12.6 from /mnt/c/Users/valarnin/Source/valarnin-cactbot/node_modules/ajv
   └─ Version 8.11.0 from /mnt/c/Users/valarnin/Source/valarnin-cactbot/node_modules/copy-webpack-plugin/node_modules/ajv
```

I'm not sure what can be done about large trigger files, but the `popup-text.ts` and `test_trigger.ts` cases could probably be improved?

**Comment by quisquous** [comment]
Oooh, interesting!  Thanks for the analysis.  Which `popup-text.ts` lines are those? I'm not quite sure I follow what the output is pointing at here.

**Comment by valarnin** [comment]
Sorry, I had forgotten that I had played around with fixes between when I originally ran the first trace and when I re-did the analysis for opening this issue. Updated the trace outputs above to actually match the `main` code.

It's a standard trace tree, so it starts from more generic (file-level) and works down to more specific (exact token).

So for `popup-text.ts`, it's saying type checking the entire file takes `2467ms` to compile.

Type checking the segment from `697:11 -> 697:39` took `1918ms` of that `2467ms` time. (`this.ProcessTrigger(trigger)`)

Type checking the segment from `697:31 -> 697:38` took `1915ms` of that `1918ms` time. (`trigger`)

Basically, it's saying that type checking the `trigger` parameter is taking almost 2 full seconds.

**Comment by trim21** [comment]
I did a little digging in here.

Currently (after the patch of new trigger style), the slowness of `popup-text.ts` file is beacuse when getting or setting a property of trigger (like `trigger.id`), it need to narrow the type of trigger (here it's `BaseTrigger<RaidbossData, "GameLog"> & PartialNetRegexTrigger<"GameLog"> & Partial<BaseTrigger<RaidbossData, "None"> & PartialRegexTrigger & PartialNetRegexTrigger<...>>) | ... 39 more ... | (BaseTrigger<...> & ... 2 more ...`) to anyone of them, and it's really slow (tsc even can't show the full list!).

And it's not hard to solve, we create a new variable `const processed: ProcessedTrigger = {}` and copy all fields, and set property to this new processed trigger, and `orderedTriggers.push(processed);`

roughly like this:

```diff
           const id = trigger.id;

+          if (trigger.disabled === true)
+            continue;
+
           if (!isRegexTrigger(trigger) && !isNetRegexTrigger(trigger)) {
             console.error(`Trigger ${id}: has no regex property specified`);
             continue;
           }

-          this.ProcessTrigger(trigger);
-
           let found = false;
+          const processed: ProcessedTrigger = {
+            id: trigger.id,
+            filename: trigger.filename,
+            condition: trigger.condition,
+            preRun: trigger.preRun,
+            delaySeconds: trigger.delaySeconds,
+            durationSeconds: trigger.durationSeconds,
+            suppressSeconds: trigger.suppressSeconds,
+            promise: trigger.promise,
+            sound: trigger.sound,
+            soundVolume: trigger.soundVolume,
+            response: trigger.response,
+            alarmText: trigger.alarmText,
+            alertText: trigger.alertText,
+            infoText: trigger.infoText,
+            tts: trigger.tts,
+            run: trigger.run,
+            output: TriggerOutputProxy.makeOutput(
+              trigger,
+              this.options.DisplayLanguage,
+              this.options.PerTriggerAutoConfig,
+            ),
+          };

           const triggerObject: { [key: string]: unknown } = trigger;

@@ -739,13 +765,11 @@ export class PopupText {
             const defaultRegex = trigger.regex;
             const localeRegex = triggerObject[regexParserLang];
             if (localeRegex instanceof RegExp) {
-              trigger.localRegex = Regexes.parse(localeRegex);
-              orderedTriggers.push(trigger);
+              processed.localRegex = Regexes.parse(localeRegex);
               found = true;
             } else if (defaultRegex) {
               const trans = translateRegex(defaultRegex, this.parserLang, set.timelineReplace);
-              trigger.localRegex = Regexes.parse(trans);
-              orderedTriggers.push(trigger);
+              processed.localRegex = Regexes.parse(trans);
               found = true;
             }
           }
@@ -762,12 +786,11 @@ export class PopupText {
               trigger.localNetRegex = Regexes.parse(localeNetRegex);
               orderedTriggers.push(trigger);
               found = true;
-            } else if (defaultNetRegex) {
+            } else {
               // simple netRegex trigger, need to build netRegex and translate
               if (defaultNetRegex instanceof RegExp) {
                 const trans = translateRegex(defaultNetRegex, this.parserLang, set.timelineReplace);
-                trigger.localNetRegex = Regexes.parse(trans);
-                orderedTriggers.push(trigger);
+                processed.localNetRegex = Regexes.parse(trans);
                 found = true;
               } else {
                 if (trigger.type === undefined) {
@@ -779,17 +802,18 @@ export class PopupText {
                   trigger.type,
                   translateRegexBuildParam(defaultNetRegex, this.parserLang, set.timelineReplace),
                 );
-                trigger.localNetRegex = Regexes.parse(re);
-                orderedTriggers.push(trigger);
+                processed.localNetRegex = Regexes.parse(re);
                 found = true;
               }
             }
           }

           if (!found) {
-            console.error('Trigger ' + trigger.id + ': missing regex and netRegex');
+            console.error(`Trigger ${processed.id ?? ''}: missing regex and netRegex`);
             continue;
           }
+
+          orderedTriggers.push(processed);
         }
```

But it only take 1s in my all 38s building, the most of time is spending on the trigger files , checking the type of `TriggerSet<Data>`, and I didn't find a solution for this.
